lib/puppet/state_machine.rb

```diff
@@ -0,0 +1,197 @@
+    #
+    # ### Namespacing
+    #
+    # When a state is namespaced, it adds the given namespace to the state name and transitions.
+    # Given a machine like the following:
+    #
+    #     Puppet::StateMachine.build("trivial") do |m|
+    #       m.start_state(:start)
+    #       m.state(:start,
+    #         action: { Puppet.notice("trivial") },
+    #         final: true
+    #       end
+    #     end
+    #
+    # When the machine is composed with the namespace `:ns1`, the `:start`: state name becomes
+    # `[:ns1, :start]`. If the machine is namespaced again with `:ns`, the state name becomes
+    # `[:ns2, :ns1, :start]`.
+    #
+    # ### Final states
+    #
+    # States that were marked as final are converted to intermediate states so that transitions
+    # can be added between states. Converted states receive a new event
+    #
+    #
+    # @example Creating two state machines and composing them in a chain
+    #   machine1 = Puppet::StateMachine.build("Machine 1") do |m|
+    #     m.state(:m1_start,
+    #       action: -> { Puppet.notice("Starting machine 1") },
+    #       event: ->(result) { :started },
+    #       transitions: {started: :m1_final})
+    #
+    #     m.state(:m1_final,
+    #       action: -> { Puppet.notice("Finishing machine 1") },
+    #       type: :final)
+    #   end
+    #
+    #   machine2 = Puppet::StateMachine.build do |m|
+    #     m.state(:m2_start,
+    #       action: -> { Puppet.notice("Starting machine 2") },
+    #       event: ->(result) { :started },
+    #       transitions: {started: :m2_final})
+    #
+    #     m.state(:m2_final,
+    #       action: -> { Puppet.notice("Finishing machine 2") },
+    #       type: :final)
+    #   end
+    #
+    #   composed = Puppet::StateMachine.compose("Composed machine",
+    #   {
+    #     m1: machine1,
+    #     m2: machine2
+    #   }) do |m|
+    #     m.start_state(:composed_start)
+    #
+    #     # Define new start state that transitions to m1
+    #     m.state(:composed_start,
+    #       action: -> { Puppet.notice("Starting composed machine") },
+    #       event: ->(result) { :started },
+    #       transitions: {started: [:m1, :m1_start]})
+    #
+    #     # Define a new final state that m2 will transition to when finished
+    #     m.state(:composed_final,
+    #       type: :final,
+    #       action: -> { Puppet.notice("Finishing composed machine") })
+    #
+    #     # Transition from m1 when finished to m2 start
+    #     m.transition(
+    #       source: [:m1, :m1_final],
+    #       target: [:m2, :m2_start],
+    #       on: Puppet::StateMachine::FINAL)
+    #
+    #     # Transition from m2 final state to composed machine final state
+    #     m.transition(
+    #       source: [:m2, :m2_final],
+    #       target: :composed_final,
+    #       on: Puppet::StateMachine::FINAL)
+    #
+    #   end
+    #
+    def self.compose(name, machines, &block)
+      MachineBuilder.new(name).compose(machines, &block)
```

Comment:
Perhaps a stylistic thing but maybe this would be more maintainable with fully qualified namespaces, ie `Puppet::StateMachine::MachineBuilder.new` ? maybe that's too verbose